python/ray/data/_internal/execution/operators/actor_pool_submitter.py

```diff
@@ -0,0 +1,172 @@
+        block = builder.build()
+        block_metadata = BlockAccessor.for_block(block).get_metadata([], None)
+        block_metadata.exec_stats = stats.build()
+        return block, block_metadata
+
+
+class ActorPool:
+    """A pool of actors for map task execution.
+
+    This class is in charge of tracking the number of in-flight tasks per actor,
+    providing the least heavily loaded actor to the task submitter, and killing idle
+    actors when the task submitter is done submitting work to the pool.
+    """
+
+    def __init__(self):
+        # Number of tasks in flight per actor.
+        self._tasks_in_flight: Dict[ray.actor.ActorHandle, int] = {}
+        # Whether actors that become idle should be eagerly killed. This is False until
+        # the first call to kill_idle_actors().
+        self._should_kill_idle_actors = False
+
+    def add_actor(self, actor: ray.actor.ActorHandle):
+        """Adds an actor to the pool."""
+        self._tasks_in_flight[actor] = 0
+
+    def pick_actor(self) -> ray.actor.ActorHandle:
+        """Provides the least heavily loaded actor in the pool for task submission."""
+        actor = min(
+            self._tasks_in_flight.keys(), key=lambda actor: self._tasks_in_flight[actor]
+        )
+        self._tasks_in_flight[actor] += 1
+        return actor
+
+    def return_actor(self, actor: ray.actor.ActorHandle):
+        """Returns the provided actor to the pool."""
+        self._tasks_in_flight[actor] -= 1
+        if self._should_kill_idle_actors and self._tasks_in_flight[actor] == 0:
+            self._kill_actor(actor)
+
+    def has_actor(self, actor: ray.actor.ActorHandle) -> bool:
+        """Whether the provided actor is in this pool."""
+        return actor in self._tasks_in_flight
+
+    def get_tasks_in_flight(self, actor: ray.actor.ActorHandle) -> int:
+        """The number of tasks in flight for the provided actor."""
+        return self._tasks_in_flight[actor]
+
+    @property
+    def num_actors(self) -> int:
+        """The number of actors in the pool."""
+        return len(self._tasks_in_flight)
+
+    def kill_idle_actors(self):
+        """Kills all currently idle actors.
+
+        This is called once the task submitter is done submitting work to the pool.
+        """
+        idle_actors = [
+            actor
+            for actor, tasks_in_flight in self._tasks_in_flight.items()
+            if tasks_in_flight == 0
+        ]
+        for actor in idle_actors:
+            self._kill_actor(actor)
+
+    def kill_future_idle_actors(self):
+        """Ensure that all actors that become idle in the future will be eagerly killed.
+
+        This is called once the task submitter is done submitting work to the pool.
+        """
+        self._should_kill_idle_actors = True
+
+    def _kill_actor(self, actor: ray.actor.ActorHandle):
+        """Kill the provided actor and remove it from the pool."""
+        ray.kill(actor)
+        del self._tasks_in_flight[actor]
```

Comment:
I think we also need to kill all actors on shutdown (seems legacy compute does it as follows):

```
        except Exception as e:
            try:
                for worker in workers:
                    ray.kill(worker)
            except Exception as err:
                logger.exception(f"Error killing workers: {err}")
            finally:
                raise e from None
```

Reply:
@ericl Ah yep, this actor pool currently kills actors once (1) no more work is being submitted, and (2) the actor becomes idle, but that won't cover killing actors when an error is raised; I'll had a submitter shutdown before I thought bases were covered by idle scale down, so I'll add that back!